src/Microsoft.Restier.Core/Conventions/ConventionBasedApiModelBuilder.cs

```diff
@@ -56,130 +46,14 @@ private ConventionBasedApiModelBuilder(Type targetType)
-            }
-
-            if (relevantType == null)
-            {
-                var singletonProperty = this.singletonProperties.SingleOrDefault(p => p.Name == name);
-                if (singletonProperty != null)
-                {
-                    relevantType = singletonProperty.PropertyType;
-                }
-            }
-
-            if (relevantType != null)
-            {
-                return true;
-            }
-
-            var innerHandler = ((IDelegateHookHandler<IModelMapper>)this).InnerHandler;
-            return innerHandler != null && innerHandler.TryGetRelevantType(context, name, out relevantType);
-        }
+            // The model builder must maintain a singleton life time, for holding states and being injected into
+            // some other services.
+            builder.AddInstance(new ConventionBasedApiModelBuilder(targetType));
 
-        /// <inheritdoc/>
-        public bool TryGetRelevantType(
-            ApiContext context,
-            string namespaceName,
-            string name,
-            out Type relevantType)
-        {
-            relevantType = null;
-            return false;
-        }
-
-        /// <inheritdoc/>
-        public Expression Expand(QueryExpressionContext context)
-        {
-            IQueryable result = GetEntitySetQuery(context);
-
-            if (result != null)
-            {
-                // Only Expand to expression of method call on ApiData class
-                var methodCall = result.Expression as MethodCallExpression;
-                if (methodCall != null)
-                {
-                    var method = methodCall.Method;
-                    if (method.DeclaringType == typeof(ApiData) && method.Name != "Value")
-                    {
-                        return null;
-                    }
-                    else
-                    {
-                        return result.Expression;
-                    }
-                }
-
-                return null;
-            }
-
-            return null;
-        }
-
-        /// <inheritdoc/>
-        public Expression Source(QueryExpressionContext context, bool embedded)
-        {
-            var innerHandler = ((IDelegateHookHandler<IQueryExpressionSourcer>)this).InnerHandler;
-            if (innerHandler != null)
-            {
-                var result = innerHandler.Source(context, embedded);
-                if (result != null)
-                {
-                    return result;
-                }
-            }
-
-            var query = GetEntitySetQuery(context) ?? GetSingletonQuery(context);
-            if (query != null)
-            {
-                return Expression.Constant(query);
-            }
-
-            return null;
+            builder.ChainPrevious<IModelBuilder, ModelBuilder>();
+            builder.ChainPrevious<IModelMapper, ModelMapper>();
+            builder.CutoffPrevious<IQueryExpressionExpander, QueryExpressionExpander>();
```

Comment:
I think so. All convention-based or default hook handlers should be the inner most.